### ISSUE TITLE
Cannot proceed to "Extract object captions and build scene graphs" with streamlined version due to mask dimension mismatch

### ISSUE THREAD

**Issue by eeaurora** [concept-graphs/concept-graphs#28]
Thanks for your great work and activate updating.

After generation of `output.pkl.gz` with streamlined detection and mapping (or realtime mapping) version, I feed it to [Extract object captions](https://github.com/concept-graphs/concept-graphs/tree/ali-dev?tab=readme-ov-file#extract-object-captions-and-build-scene-graphs) with `extract-node-captions mode` to generate caption with LLava. But something is wrong with **mask shape** saved in `output.pkl.gz`

The error is due to `Initial shape mismatch: Image shape (680, 1200, 3) != Mask shape (1200,)  `  [here](https://github.com/concept-graphs/concept-graphs/blob/002d6f47b6d79be884c54b04cdbe0144f4753ba0/conceptgraph/scenegraph/build_scenegraph_cfslam.py#L330), [here](https://github.com/concept-graphs/concept-graphs/blob/002d6f47b6d79be884c54b04cdbe0144f4753ba0/conceptgraph/scenegraph/build_scenegraph_cfslam.py#L166)

I simply `print("obj["mask"]  shape: " , np.array(obj["mask"]).shape)`  in [this line](https://github.com/concept-graphs/concept-graphs/blob/002d6f47b6d79be884c54b04cdbe0144f4753ba0/conceptgraph/scenegraph/build_scenegraph_cfslam.py#L326),  the shape is [680, 1200]. 

But the `output.pkl.gz` generated with the original version can run successfully, and the shape of `obj["mask"]` is like [xxx, 680, 1200].

The mask dimension is 3 in original version, but it is 2 in streamlined (or realtime) version.  Maybe something is missing when saving mask information in streamlined version. 

**Can you please fix it to compatible with `build_scenegraph_cfslam.py` to proceed to Extract object captions and build scene graphs? Thank you.**

```
SCENE_NAME=room0
PKL_FILENAME=pcd_s_mapping_stride50_1.pkl.gz  # Change this to the actual output file name of the pkl.gz file

python scenegraph/build_scenegraph_cfslam.py \
    --mode extract-node-captions \
    --cachedir ${REPLICA_ROOT}/${SCENE_NAME}/exps/sg_cache \
    --mapfile ${REPLICA_ROOT}/${SCENE_NAME}/exps/s_detections_stride50_2/${PKL_FILENAME} \
    --class_names_file /share/Project/AI/LLM/LLMGraphServer/Datasets/Replica/room0_origin/gsa_classes_ram_withbg_allclasses.json

args.masking_option: none
Loaded 74 objects
['door', 'curtain', 'sink', 'footrest', 'table', 'couch', 'boat', 'swivel chair', 'cat', 'ceiling', 'daybed', 'anchor', 'hang', 'drawer', 'blind', 'other item', 'den', 'cabinet', 'bay window', 'side table', 'window', 'hassock', 'blanket', 'lead to', 'lamp', 'wall', 'stool', 'chair', 'floor', 'vanity', 'dresser', 'furniture', 'fill', 'sculpture', 'ceiling fan', 'nightstand', 'white', 'armchair', 'glass table', 'round table', 'pillow', 'pad']
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████| 2/2 [00:04<00:00,  2.09s/it]
LLaVA chat initialized...
  0%|                                                                                               | 0/74 [00:00<?, ?it/sInitial shape mismatch: Image shape (680, 1200, 3) != Mask shape (1200,)                              | 0/3 [00:00<?, ?it/s]
  0%|                                                                                                | 0/3 [00:00<?, ?it/s]
  0%|                                                                                               | 0/74 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/share/Project/AI/LLM/LLMGraphServer/concept-graphs/conceptgraph/scenegraph/build_scenegraph_cfslam.py", line 975, in <module>
    main()
  File "/share/Project/AI/LLM/LLMGraphServer/concept-graphs/conceptgraph/scenegraph/build_scenegraph_cfslam.py", line 960, in main
    extract_node_captions(args)
  File "/share/Project/AI/LLM/LLMGraphServer/concept-graphs/conceptgraph/scenegraph/build_scenegraph_cfslam.py", line 358, in extract_node_captions
    _w, _h = image_crop.size
AttributeError: 'NoneType' object has no attribute 'size'
```
- mask info of streamlined version
<img width="458" alt="image" src="https://github.com/concept-graphs/concept-graphs/assets/163414162/f440b4cd-e37b-4f79-b64b-cfd997f850ea">

- mask info of original version
<img width="497" alt="image" src="https://github.com/concept-graphs/concept-graphs/assets/163414162/241cc577-753f-4d43-b94b-379352592e23">

**Comment by alik-git** [comment]
am looking into this, thanks for bringing it up.

**Comment by eeaurora** [comment]
Thank you for your reply.

If I understand correctly, it is not because "the mask which consists of boolean values in a single channel is not compatible with the image that consists of 3 channels".  

This [line](https://github.com/concept-graphs/concept-graphs/blob/002d6f47b6d79be884c54b04cdbe0144f4753ba0/conceptgraph/scenegraph/build_scenegraph_cfslam.py#L165) is to check  `if image.shape[:2] != mask.shape:`  , instead of `if image.shape != mask.shape:`, so it works well if image shape is [680,1200,3] and mask shape is [680, 1200]. It indicates that before [here](https://github.com/concept-graphs/concept-graphs/blob/002d6f47b6d79be884c54b04cdbe0144f4753ba0/conceptgraph/scenegraph/build_scenegraph_cfslam.py#L330),  we are choosing a mask from a bunch of masks by `mask = obj["mask"][idx_det]`, so the shape of `obj["mask"]` should be like [10, 680,1200]  (10 is for instance).

For example, we detect a book in 10 images (also corresponding to 10 masks of book), and [this line](https://github.com/concept-graphs/concept-graphs/blob/002d6f47b6d79be884c54b04cdbe0144f4753ba0/conceptgraph/scenegraph/build_scenegraph_cfslam.py#L325) is to choose a mask from 10 masks, so the shape of obj["mask"] is like [10, 680,1200], But now the shape of obj["mask"] is [680,1200], that is why I think mask information is missing in current `output.pkl.gz`

---

> This [for loop](https://github.com/concept-graphs/concept-graphs/blob/002d6f47b6d79be884c54b04cdbe0144f4753ba0/conceptgraph/scenegraph/build_scenegraph_cfslam.py#L300) is to choose a specific object and extract its 10 most confident index in `idx_most_conf`, this  [for loop](https://github.com/concept-graphs/concept-graphs/blob/002d6f47b6d79be884c54b04cdbe0144f4753ba0/conceptgraph/scenegraph/build_scenegraph_cfslam.py#L318)  is to extract corresponding image, xyxy, mask from `obj`, so there should be some masks for a specific object, instead of only one mask.
> 
> 
> 
> ```
>     for idx_obj, obj in tqdm(enumerate(scene_map), total=len(scene_map)):
>         conf = obj["conf"]
>         conf = np.array(conf)
>         idx_most_conf = np.argsort(conf)[::-1]
> 
>         features = []
>         captions = []
>         low_confidences = []
>         
>         image_list = []
>         caption_list = []
>         confidences_list = []
>         low_confidences_list = []
>         mask_list = []  # New list for masks
>         if len(idx_most_conf) < 2:
>             continue 
>         idx_most_conf = idx_most_conf[:args.max_detections_per_object]
> 
>         for idx_det in tqdm(idx_most_conf):
>             # image = Image.open(correct_path).convert("RGB")
>             image = Image.open(obj["color_path"][idx_det]).convert("RGB")
>             xyxy = obj["xyxy"][idx_det]
>             class_id = obj["class_id"][idx_det]
>             class_name = class_names[class_id]
>             # Retrieve and crop mask
>             mask = obj["mask"][idx_det]
> ```

**Comment by alik-git** [comment]
Hey @eeaurora, thanks for your patience, this issue should be fixed now by @sachaMorin in PR #33 

Thanks everyone!
